src/schema/__tests__/FroidSchema.test.ts

```diff
@@ -0,0 +1,1690 @@
+  it('omits interface object', () => {
+    const productSchema = gql`
+      type Product @key(fields: "upc") {
+        upc: String!
+      }
+
+      type PrintedMedia @interfaceObject @key(fields: "mediaId") {
+        mediaId: Int!
+      }
+    `;
+    const subgraphs = new Map();
+    subgraphs.set('product-subgraph', productSchema);
+
+    const actual = generateSchema({
+      subgraphs,
+      froidSubgraphName: 'relay-subgraph',
+    });
+
+    expect(actual).toEqual(
+      // prettier-ignore
+      gql`
+      extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["@key", "@tag", "@external"])
+
+      type Query {
+        node(id: ID!): Node
+      }
+
+      interface Node {
+        id: ID!
+      }
+
+      type Product implements Node @key(fields: "upc") {
+        id: ID!
+        upc: String!
+      }
+    `
+    );
+  });
+
+  it('ignores @key(fields: "id") directives', () => {
+    const productSchema = gql`
+      type Query {
+        topProducts(first: Int = 5): [Product]
+      }
+
+      type Product @key(fields: "upc") {
+        upc: String!
+        name: String
+        price: Int
+        weight: Int
+      }
+
+      type Brand @key(fields: "id") {
+        id: ID!
+        name: String
+      }
+    `;
+    const subgraphs = new Map();
+    subgraphs.set('product-subgraph', productSchema);
+
+    const actual = generateSchema({
+      subgraphs,
+      froidSubgraphName: 'relay-subgraph',
+    });
+
+    expect(actual).toEqual(
+      // prettier-ignore
+      gql`
+        extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["@key", "@tag", "@external"])
+
+        type Query {
+          node(id: ID!): Node
+        }
+
+        interface Node {
+          id: ID!
+        }
+
+        type Product implements Node @key(fields: "upc") {
+          id: ID!
+          upc: String!
+        }
+      `
+    );
+  });
+  it('does not propagate miscellaneous directives to the generated id field', () => {
+    const productSchema = gql`
+      type Query {
+        topProducts(first: Int = 5): [Product]
+      }
+
+      type Product @key(fields: "upc") {
+        upc: String! @someDirective
+        weight: Int
+      }
+    `;
+    const subgraphs = new Map();
+    subgraphs.set('product-subgraph', productSchema);
+
+    const actual = generateSchema({
+      subgraphs,
+      froidSubgraphName: 'relay-subgraph',
+    });
+
+    expect(actual).toEqual(
+      // prettier-ignore
+      gql`
+        extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["@key", "@tag", "@external"])
+
+        type Query {
+          node(id: ID!): Node
+        }
+
+        interface Node {
+          id: ID!
+        }
+
+        type Product implements Node @key(fields: "upc") {
+          id: ID!
+          upc: String!
+        }
+      `
+    );
+  });
+
+  it('defaults to generating valid schema using the first key found regardless of complexity', () => {
```

Comment:
description does not look to correspond

Reply:
The description is accurate. It's picking the first `Product` key to use in the FROID schema. It is, however, also injecting `__typename`. But otherwise the key is the same.

Reply:
so my understanding is that noderelay key will be an aggregation of all the fields that can be found in all the possible keys

```
# subgraph
type Foo
@key(fields: "a")
@key(fields: "b")
{
...
}

#Froid
type Foo implements Node
@key(fields: "a b")
{
...
}
```

but here the test is saying we pick the first key?
How do we say to froid -> use the aggregated set of keys?

if this test is valid, don't set the first key to be the aggregate list nor the most complex, nor the most simple
`upc sku` should be used

Reply:
I don't see tests like the example 👆 , or examples you describe in the PR description. Do I miss something?

Reply:
The tests do replicate what's described in the PR.

A key has to be selected for the initial entity in order for it to exist. This is the first key. Most entities are not used in a complex key, therefore just the first key is used.

We don't automatically aggregate all keys. We only aggregate keys that are used by the initial entity (the first key) or by another entity. This way we are not senselessly aggregating keys.

Reply:
Arguably, we could throw away the initial key picked if all entities that incorporate that entity into their complex keys are using another key. That might make sense, as it would result in potentially fewer compound keys.

I'm going to take a look at doing that in my PR.